src-self-hosted/link/Wasm.zig

```diff
@@ -0,0 +1,445 @@
+    const code_id = 10;
+    const data_id = 11;
+};
+
+pub const base_tag = link.File.Tag.wasm;
+
+pub const FnData = struct {
+    funcidx: u32,
+    typeidx: u32,
+};
+
+base: link.File,
+
+types: Types,
+funcs: Funcs,
+exports: Exports,
+
+/// Array over the section structs used in the various sections above to
+/// allow iteration when shifting sections to make space.
+/// TODO: this should eventually be size 11 when we use all the sections.
+sections: [4]*Section,
+
+pub fn openPath(allocator: *Allocator, dir: fs.Dir, sub_path: []const u8, options: link.Options) !*link.File {
+    assert(options.object_format == .wasm);
+
+    // TODO: read the file and keep vaild parts instead of truncating
+    const file = try dir.createFile(sub_path, .{ .truncate = true, .read = true });
+    errdefer file.close();
+
+    const wasm = try allocator.create(Wasm);
+    errdefer allocator.destroy(wasm);
+
+    try file.writeAll(&(spec.magic ++ spec.version));
+
+    wasm.base = .{
+        .tag = .wasm,
+        .options = options,
+        .file = file,
+        .allocator = allocator,
+    };
+
+    // TODO: this should vary depending on the section and be less arbitrary
+    const size = 1024;
+    const offset = @sizeOf(@TypeOf(spec.magic ++ spec.version));
+
+    wasm.types = try Types.init(file, offset, size);
+    wasm.funcs = try Funcs.init(file, offset + size, size, offset + 3 * size, size);
+    wasm.exports = try Exports.init(file, offset + 2 * size, size);
+    try file.setEndPos(offset + 4 * size);
+
+    wasm.sections = [_]*Section{
+        &wasm.types.typesec.section,
+        &wasm.funcs.funcsec,
+        &wasm.exports.exportsec,
+        &wasm.funcs.codesec.section,
+    };
+
+    return &wasm.base;
+}
+
+pub fn deinit(self: *Wasm) void {
+    if (self.base.file) |f| f.close();
+    self.types.deinit();
+    self.funcs.deinit();
+}
+
+pub fn updateDecl(self: *Wasm, module: *Module, decl: *Module.Decl) !void {
+    if (decl.typed_value.most_recent.typed_value.ty.zigTypeTag() != .Fn)
+        return error.TODOImplementNonFnDeclsForWasm;
+
+    if (decl.fn_link.wasm) |fn_data| {
+        self.types.free(fn_data.typeidx);
+        self.funcs.free(fn_data.funcidx);
+    }
+
+    var buf = std.ArrayList(u8).init(self.base.allocator);
+    defer buf.deinit();
+
+    try codegen.genFunctype(&buf, decl);
+    const typeidx = try self.types.new(buf.items);
+    buf.items.len = 0;
+
+    try codegen.genCode(&buf, decl);
+    const funcidx = try self.funcs.new(typeidx, buf.items);
+
+    decl.fn_link.wasm = .{ .typeidx = typeidx, .funcidx = funcidx };
+
+    try self.exports.writeAll(module);
+}
+
+pub fn updateDeclExports(
+    self: *Wasm,
+    module: *Module,
+    decl: *const Module.Decl,
+    exports: []const *Module.Export,
+) !void {
+    // TODO: updateDeclExports() may currently be called before updateDecl,
+    // presumably due to a bug. For now just rely on the following call
+    // being made in updateDecl().
```

Comment:
I'll look into this

Reply:
I think this is the only change that needs to happen regarding this:

```diff
diff --git a/src-self-hosted/link.zig b/src-self-hosted/link.zig
index a835cc6b7..e89210763 100644
--- a/src-self-hosted/link.zig
+++ b/src-self-hosted/link.zig
@@ -186,7 +186,8 @@ pub const File = struct {
         };
     }
 
-    /// Must be called only after a successful call to `updateDecl`.
+    /// Can be called any time before or after the call to `updateDecl`. Same
+    /// with `allocateDeclIndexes`.
     pub fn updateDeclExports(
         base: *File,
         module: *Module,

```

Happy to discuss other possible ways of doing things